### DISCUSSION TITLE
Handling packages exports

### DISCUSSION THREAD

**Discussion by pepew-le-boss** [vercel/turborepo#6047] [Help] [answered]
I've been working on a NextJS project for almost a year now. Because we are about to work on another NextJS project that will share a lot of code with the first one, I've decided to setup a monorepo since it seems to be a very good solution for this use case.

I've taken some of my code and split it inside the packages folder, meaning that now it contains all those folders: `/components`, `/utils`, `/graphql`, `/data` and so on.

Basically what I've done inside each folder is the following: 
```
/components
-- package.json
-- tsconfig.json
-- /src
---- all my previous code
```

My problem is that when I try to use, let's say, a `utils` function in my `components` folder I don't get any auto-import functionality.
I can still specify the path by hand and it would work but the auto-import never suggest anything.
For example I can do this by hand and it works: 
`import { tailwindBreakPoints } from "@company/utils/src/lib/tailwind/getTailwindConfig"`.
Howerver if I comment out this line and try to auto-import `tailwindBreakPoints`, I'll have no suggestion.

I understand that if I want to that correctly I need to add an `index.ts` in each package and export all of my files and functions from there.
But I would rather not specify any exports by hand for 3 reasons:
- I've a ton of files to export (meaning spending days just writing exports)
- In the future for each component I write, each file I create, each folder I add, I will need to add an export (that's quite annoying)
- we are only 2 devs working on this codebase so we don't need a full on library kind of functionality (accessing the src code directly is fine with us).

So how can I fix the typescript so the auto-imports would work ?

Example of a `package.json` inside a `package`:
``` json
{
  "name": "@company/components",
  "version": "0.0.0",
  "license": "MIT",
  "dependencies": {
    "@company/utils": "*",
    "@company/gql-generated": "*"
  },
  "devDependencies": {
    "@company/tsconfig": "*"
  }
}
```

Example of a `tsconfig.json` inside a `package`:
``` json
{
  "extends": "@company/tsconfig/pckg.json",
  "include": ["."],
  "exclude": ["dist", "build", "node_modules"],
  "compilerOptions": {
    "baseUrl": ".",
    "paths": {
      "@/components/*": ["./src/*"],
    }
  }
}
```

The `pckg.json`:
``` json
{
  "compilerOptions": {
    "target": "esnext",
    "lib": [
      "dom",
      "dom.iterable",
      "esnext"
    ],
    "allowJs": true,
    "skipLibCheck": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "noEmit": true,
    "esModuleInterop": true,
    "module": "esnext",
    "moduleResolution": "nodenext",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "jsx": "preserve",
    "incremental": true
  },
  "exclude": [
    "node_modules"
  ]
}
```

**Comment by anthonyshew** [comment, marked as answer]
Check out our updated examples in [the examples directory](https://github.com/vercel/turbo/tree/main/examples/basic). Auto-importing works there because of the `exports` key in `./packages/ui/package.json`.
